### PR TITLE
Add style rule for mandatory braces for control flow statements

### PR DESCRIPTION
**Pull request by jkaan**
This adds the rule that was talked about in #365

### CODE REVIEW

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/optional/MandatoryBracesIfStatements.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+		super.visitIfExpression(expression)
+	}
+
+	private fun hasNewLine(expression: KtIfExpression): Boolean =
+			expression.rightParenthesis?.siblings(true, false)
+					?.filter { it is PsiWhiteSpace }
+					?.firstOrNull { (it as PsiWhiteSpace).textContains('\n') } != null
+
+	private fun isNotBlockExpression(expression: KtIfExpression): Boolean =
+			expression.then?.node?.elementType != KtNodeTypes.BLOCK
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/optional/MandatoryBracesIfStatements.kt]: you can check `then` if it's an instance of `KtBlockExpression`

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/optional/MandatoryBracesIfStatements.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+		if (isNotBlockExpression(expression) && hasNewLine(expression)) {
+			report(CodeSmell(issue, Entity.from(expression), message = ""))
+		}
+		super.visitIfExpression(expression)
+	}
+
+	private fun hasNewLine(expression: KtIfExpression): Boolean =
+			expression.rightParenthesis?.siblings(true, false)
+					?.filter { it is PsiWhiteSpace }
+					?.firstOrNull { (it as PsiWhiteSpace).textContains('\n') } != null
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/optional/MandatoryBracesIfStatements.kt]: Please use a safe cast here

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/optional/MandatoryBracesIfStatements.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+class MandatoryBracesIfStatements(config: Config = Config.empty) : Rule(config) {
+
+	override val issue = Issue("MandatoryBracesIfStatements", Severity.Style,
+			"Multi-line if statements detected without braces. " +
+					"The braces can be added to improve readability and prevent possible errors.",
+			Debt.FIVE_MINS)
+
+	override fun visitIfExpression(expression: KtIfExpression) {
+		if (isNotBlockExpression(expression) && hasNewLine(expression)) {
+			report(CodeSmell(issue, Entity.from(expression), message = ""))
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/optional/MandatoryBracesIfStatements.kt]: Please provide a message here. This really helps the detekt user to get a clue of the violation.

**Comment by jkaan** [comment]
@schalkms Thanks for the review! Would you mind taking another look?

I've looked at all the comments and applied them :). I added them as two separate commits but I can squash them into one once it's mergeable.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/optional/MandatoryBracesIfStatements.kt [hunk 4]
```diff
@@ -7,13 +7,13 @@ import io.gitlab.arturbosch.detekt.api.Entity
...
 import io.gitlab.arturbosch.detekt.api.Severity
-import org.jetbrains.kotlin.KtNodeTypes
 import org.jetbrains.kotlin.com.intellij.psi.PsiWhiteSpace
+import org.jetbrains.kotlin.psi.KtBlockExpression
 import org.jetbrains.kotlin.psi.KtIfExpression
 import org.jetbrains.kotlin.psi.psiUtil.siblings
 
 /**
- * This rule detects multi-line `if` statements which does have braces.
+ * This rule detects multi-line `if` statements which does not have braces.
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/optional/MandatoryBracesIfStatements.kt]: `do not`

File: detekt-rules/src/test/resources/cases/MandatoryBracesIfStatementsNegative.kt [hunk 5]
```diff
@@ -14,4 +14,8 @@ fun mandatoryBracesIfStatementNegative() {
 	if (true) println()
 
 	if (true) { println() }
+
+	if (true) println() else println()
```
Comment (schalkms) [inline detekt-rules/src/test/resources/cases/MandatoryBracesIfStatementsNegative.kt]: I would also add some test cases for
```Kotlin
if (true) {
  //
} else 
  print()

if (...)
{
}
```

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/optional/MandatoryBracesIfStatements.kt [hunk 6]
```diff
@@ -37,16 +37,18 @@ class MandatoryBracesIfStatements(config: Config = Config.empty) : Rule(config)
...
 		}
 		super.visitIfExpression(expression)
 	}
 
 	private fun hasNewLine(expression: KtIfExpression): Boolean =
 			expression.rightParenthesis?.siblings(true, false)
-					?.filter { it is PsiWhiteSpace }
-					?.firstOrNull { (it as PsiWhiteSpace).textContains('\n') } != null
+					?.filterIsInstance<PsiWhiteSpace>()
+					?.firstOrNull { it.textContains('\n') } != null
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/optional/MandatoryBracesIfStatements.kt]: That improves the readability!

File: detekt-rules/src/test/resources/cases/MandatoryBracesIfStatementsNegative.kt (hunk 5 above)
Comment (jkaan) [reply detekt-rules/src/test/resources/cases/MandatoryBracesIfStatementsNegative.kt]: @schalkms I did at a test case for 

    if (...)
    {
    }

I expected this to not trigger this rule, would you think the same?

What about the this one?

    if (true) {
        //
    } else
        print()

This should probably trigger the fact that there are no braces around for the `else` statement, right?

**Comment by arturbosch** [comment]
Hey, thanks for the contribution!
I think you should handle the else-branch too.
As @schalkms mentioned, this does not get detected:
```
if (true) {
  //
} else 
  print()
```

**Comment by jkaan** [comment]
@arturbosch Yes, I totally agree. I'll add it :) thanks

**Pull request merged**